src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpRetry.cs

```diff
@@ -0,0 +1,248 @@
+
+        // TODO: Consider introducing a fixed max number of retries (e.g. max 5 retries).
+        // The spec does not specify a max number of retries, but it may be bad to not cap the number of attempts.
+        // Without a max number of retry attempts, retries would continue until the deadline.
+        // Maybe this is ok? However, it may lead to an unexpected behavioral change. For example:
+        //    1) When using a batch processor, a longer delay due to repeated
+        //       retries up to the deadline may lead to a higher chance that the queue will be exhausted.
+        //    2) When using the simple processor, a longer delay due to repeated
+        //       retries up to the deadline will lead to a prolonged blocking call.
+        // if (attemptCount >= MaxAttempts)
+        // {
+        //     return false
+        // }
+
+        if (IsDeadlineExceeded(deadline))
+        {
+            return false;
+        }
+
+        var throttleDelay = throttleGetter(statusCode, carrier);
+        var retryable = isRetryable(statusCode, throttleDelay.HasValue);
+        if (!retryable)
+        {
+            return false;
+        }
+
+        var delayDuration = throttleDelay.HasValue
+            ? throttleDelay.Value
+            : TimeSpan.FromMilliseconds(GetRandomNumber(0, nextRetryDelayMilliseconds));
+
+        if (deadline.HasValue && IsDeadlineExceeded(deadline + delayDuration))
+        {
+            return false;
+        }
+
+        if (throttleDelay.HasValue)
+        {
+            try
+            {
+                nextRetryDelayMilliseconds = Convert.ToInt32(throttleDelay.Value.TotalMilliseconds);
+            }
+            catch (OverflowException)
+            {
+                nextRetryDelayMilliseconds = MaxBackoffMilliseconds;
+            }
+        }
+
+        nextRetryDelayMilliseconds = CalculateNextRetryDelay(nextRetryDelayMilliseconds);
+        retryResult = new RetryResult(throttleDelay.HasValue, delayDuration, nextRetryDelayMilliseconds);
+        return true;
+    }
+
+    private static bool IsDeadlineExceeded(DateTime? deadline)
+    {
+        // This implementation is internal, and it is guaranteed that deadline is UTC.
+        return deadline.HasValue && deadline <= DateTime.UtcNow;
+    }
+
+    private static int CalculateNextRetryDelay(int nextRetryDelayMilliseconds)
+    {
+        var nextMilliseconds = nextRetryDelayMilliseconds * BackoffMultiplier;
+        nextMilliseconds = Math.Min(nextMilliseconds, MaxBackoffMilliseconds);
+        return Convert.ToInt32(nextMilliseconds);
+    }
+
+    private static TimeSpan? TryGetGrpcRetryDelay(StatusCode statusCode, Metadata trailers)
+    {
+        Debug.Assert(trailers != null, "trailers was null");
+
+        if (statusCode != StatusCode.ResourceExhausted && statusCode != StatusCode.Unavailable)
+        {
+            return null;
+        }
+
+        var statusDetails = trailers.Get(GrpcStatusDetailsHeader);
```

Comment:
Is every Grpc service expected to set this trailer?

Reply:
Metadata.Get returns null if the header does not exist.

https://grpc.github.io/grpc/csharp/api/Grpc.Core.Metadata.html#Grpc_Core_Metadata_Get_System_String_

Reply:
My question was more around the general availability of this header for our use. Is this left to the service implementation? Could we expect OTLP Collector to be setting this header?

Reply:
> Could we expect OTLP Collector to be setting this header?

Yes, per the spec, any OTLP receiver like the one for the collector should support this header for throttling requests. https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#otlpgrpc-throttling

Reply:
Ah, but yea I think I see the confusion... the spec does not make it obvious that it's a header by the name `grpc-status-details-bin`. This is the header the collector responds with. Similarly it's the header that New Relic's OTLP ingest responds with.

Reply:
> This is the header the collector responds with.

I think it would be good to add a code comment about this.